### PR TITLE
Simplify the Certify component and add tests (#1266)

### PR DESCRIPTION
**Pull request by WalterMoar**
*Issue #:* /bcgov/entity#1266

*Description of Certify changes:*
  - Removed the shadow property fields to simplify the component.
  - All the watchers could be removed once above simplification done.
  - Added a property for the `currentDate`, so that (1) the component isn't tied to VueX, and (2) the parent component isn't required to have a specific named variable in the store for the current date.
  - Added a `valid` event so the parent component knows if the form is complete or not. Will be returned on create so that the parent knows if the properties are valid or not.
  - Added unit tests.

*Description of AnnualReport changes:*
  - Passing in `currentDate` as a property.
  - Added a `certifyFormValid` variable - there shouldn't be any need to store any of these in the store, but leaving the others as-is for now.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

File: coops-ui/src/components/AnnualReport/Certify.vue [hunk 1]
```diff
@@ -33,83 +37,59 @@
...
+
     return val
   }
 
   // Emit an update event.
   @Emit('update:isCertified')
-  private emitIsCertified (val) {
-    return val
+  private emitIsCertified (val: boolean): boolean {
+    this.emitValid(this.certifiedBy && !!val)
```
Comment (severinbeauvais) [inline coops-ui/src/components/AnnualReport/Certify.vue]: This should check this.trimmedCertifiedBy instead.

File: coops-ui/src/components/AnnualReport/Certify.vue [hunk 2]
```diff
@@ -33,83 +37,59 @@
...
+
+    return !!val
+  }
+
+  // Emit an event indicating whether or not the form is valid.
+  @Emit('valid')
+  private emitValid (val: boolean): boolean {
+    return !!val
```
Comment (severinbeauvais) [inline coops-ui/src/components/AnnualReport/Certify.vue]: Why is !! needed here? The val passed in is already boolean, right?

File: coops-ui/src/components/AnnualReport/Certify.vue [hunk 3]
```diff
@@ -33,83 +37,59 @@
...
   }
 }
+
```
Comment (severinbeauvais) [inline coops-ui/src/components/AnnualReport/Certify.vue]: Interesting. My linter complains about whitespace below the <script> tag so I don't include one here. You included whitespace in both places. Did `npm run lint:nofix` pass?

File: coops-ui/src/components/AnnualReport/Certify.vue [hunk 4]
```diff
@@ -33,83 +37,59 @@
...
-    this.emitIsCertified(val)
+  private get trimmedCertifiedBy (): string {
+    return this.certifiedBy && this.certifiedBy.trim()
   }
 
   // Emit an update event.
   @Emit('update:certifiedBy')
-  private emitCertifiedBy (val) {
+  private emitCertifiedBy (val: string): string {
+    this.emitValid(val && val.trim() && this.isCertified)
```
Comment (severinbeauvais) [inline coops-ui/src/components/AnnualReport/Certify.vue]: In other places you use this.trimmedCertifiedBy and here you don't. I'd prefer to see consistency.

File: coops-ui/tests/unit/Certify.spec.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
+//
```
Comment (severinbeauvais) [inline coops-ui/tests/unit/Certify.spec.ts]: Is this header required? If so, please create a ticket to update all the other files. Thx!

File: coops-ui/tests/unit/Certify.spec.ts [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+import Vue from 'vue'
+import Vuetify from 'vuetify'
+
+import { mount, Wrapper } from '@vue/test-utils'
+
+import Certify from '@/components/AnnualReport/Certify.vue'
+
+Vue.use(Vuetify)
+
+// Boilerplate to prevent the complaint "[Vuetify] Unable to locate target [data-app]"
```
Comment (severinbeauvais) [inline coops-ui/tests/unit/Certify.spec.ts]: This boilerplate is not needed in every file. Did you check without it?

File: coops-ui/tests/unit/Certify.spec.ts [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+ * Returns the last event for a given name, to be used for testing event propagation in response to component changes.
+ *
+ * @param wrapper the wrapper for the component that is being tested.
+ * @param name the name of the event that is to be returned.
+ *
+ * @return the value of the last named event for the wrapper.
+ */
+function getLastEvent (wrapper: Wrapper<Certify>, name: string): any {
+  const eventsList: Array<any> = wrapper.emitted(name)
+  const events: Array<any> = eventsList[eventsList.length - 1]
```
Comment (severinbeauvais) [inline coops-ui/tests/unit/Certify.spec.ts]: What happens if an event was never emitted? :P

File: coops-ui/src/components/AnnualReport/Certify.vue (hunk 1 above)
Comment (WalterMoar) [reply coops-ui/src/components/AnnualReport/Certify.vue]: Yes! Also Indicates that there are tests missing for all-whitespace certifiedBy, so I added them.

File: coops-ui/src/components/AnnualReport/Certify.vue (hunk 2 above)
Comment (WalterMoar) [reply coops-ui/src/components/AnnualReport/Certify.vue]: Not needed, likewise with the one in emitIsCertified.

File: coops-ui/src/components/AnnualReport/Certify.vue (hunk 4 above)
Comment (WalterMoar) [reply coops-ui/src/components/AnnualReport/Certify.vue]: trimmedCertifiedBy is the computed value for certifiedBy, but in this case val != certifiedBy yes. It has to be emitted first.

File: coops-ui/tests/unit/Certify.spec.ts (hunk 5 above)
Comment (WalterMoar) [reply coops-ui/tests/unit/Certify.spec.ts]: Required? No idea. Probably a decision for the team to make.

File: coops-ui/tests/unit/Certify.spec.ts (hunk 6 above)
Comment (WalterMoar) [reply coops-ui/tests/unit/Certify.spec.ts]: I didn't check - works fine without it.

File: coops-ui/tests/unit/Certify.spec.ts (hunk 7 above)
Comment (WalterMoar) [reply coops-ui/tests/unit/Certify.spec.ts]: Good question - "it explodes" is the answer.

The tests always test for existence before calling getLastEvent,, but why not just let it fail? The resulting stack trace is enough information to find the caller and determine which event is missing.

File: coops-ui/src/components/AnnualReport/Certify.vue (hunk 3 above)
Comment (WalterMoar) [reply coops-ui/src/components/AnnualReport/Certify.vue]: I never run the linter manually, the IDE indicates when there are problems. However, running it manually does not produce a problem. The team should probably pick a lane when it comes to the IDE.

File: coops-ui/tests/unit/Certify.spec.ts (hunk 5 above)
Comment (WalterMoar) [reply coops-ui/tests/unit/Certify.spec.ts]: It is required. Opened #1291 as tech debt.

**Comment by WalterMoar** [comment]
> Looks good EXCEPT that we should discuss the copyright header as you may be advised to remove it.

We need the copyright/license. Opened #1291 as tech debt.
